### ISSUE TITLE
tensorflow2.0 detected 'xla_gpu' , but 'gpu' expected

### ISSUE THREAD

**Issue by ChuangLee** [tensorflow/tensorflow#30388]
**System information**
-  Linux Ubuntu 16.04 :
- TensorFlow installed from binary, tensorflow2.0-gpu-beta1:
- Python version 3.6.4:
- CUDA 9.1, cuDNN 7.0:
- GPU: Titan Xp.
 
tensorflow detected 'xla_gpu' , but 'gpu' expected

`train_model = keras.utils.multi_gpu_model(train_model, gpus=2)
  File "/usr/local/lib/python3.6/site-packages/tensorflow/python/keras/utils/multi_gpu_utils.py", line 182, in multi_gpu_model
    available_devices))
ValueError: To call `multi_gpu_model` with `gpus=2`, we expect the following devices to be available: ['/cpu:0', '/gpu:0', '/gpu:1']. However this machine only has: ['/cpu:0', '/xla_gpu:0', '/xla_gpu:1', '/xla_cpu:0']. Try reducing `gpus`.
`

**Comment by ziyigogogo** [comment]
In tf2.0, using [distribute_strategy](https://www.tensorflow.org/beta/guide/distribute_strategy) to utilize multiple gpus is a better way.

**Comment by ChuangLee** [comment]
> In tf2.0, using [distribute_strategy](https://www.tensorflow.org/beta/guide/distribute_strategy) to utilize multiple gpus is a better way.

Thank you , but 
if I use 
`    mirrored_strategy = tf.distribute.MirroredStrategy(
        devices=["/gpu:0", "/gpu:1"])`
then:

RuntimeError: Error copying tensor to device: /job:localhost/replica:0/task:0/device:GPU:0. /job:localhost/replica:0/task:0/device:GPU:0 unknown device.

if I use `    mirrored_strategy = tf.distribute.MirroredStrategy(
        devices=["/xla_gpu:0", "/xla_gpu:1"])`
then:
ValueError: Unknown attribute: 'xla_gpu' in '/xla_gpu:0'

**Comment by ziyigogogo** [comment]
seems that you are using a XLA compiled tf build. Not sure if it's a bug. Need some confirmation from the tf team members.

**Comment by ChuangLee** [comment]
problem solved, after update CUDA to 10.0, CUDNN to 7.4

**Comment by anilknayak** [comment]
Does this problem solved for CUDA 10.0 and CUDNN 7.4

I have installed CUDA 10.1 and CUDNN 7.5, it does not work.
```
tf.distribute.MirroredStrategy( devices=["/xla_gpu:0", "/xla_gpu:1"])
ValueError: Unknown attribute: 'xla_gpu' in '/xla_gpu:0'
``` 

same error.

**Comment by rtkaratekid** [comment]
Exact same issue for me using
CUDA 10.2
cuDNN 7.6
Tensorflow 2.0
Ubuntu 18.04
It might be a compatibility issue with the latest CUDA and cuDNN with tf? Not sure.

**Comment by mbenguig** [comment]
Any news concerning this issue ?

**Comment by rtkaratekid** [comment]
@mbenguig from the sound of all the reading I've done, the latest version of tf has some bugs in it and doesn't work properly? Honestly, until they can fix some of these issues I've given up on tensorflow roulette and am piping some of my work to pytorch to try my luck there.

**Comment by mbenguig** [comment]
Finally, you can get rid of this issue by uninstalling / reinstalling (tested on Ubuntu 18.04):

- Tensorflow 2.0
- CUDA 10.0
- cuDNN 7.6.4 (described as dedicated for CUDA 10.0)

https://www.tensorflow.org/install/source#tested_build_configurations. You will get xla devices with corresponding non xla devices.

**Comment by choco9966** [comment]
same issue
CUDA 10.2
cuDNN 7.6
Tensorflow 1.15
Ubuntu 18.04

**Comment by shamoons** [comment]
Same.
CUDA 10.1
CuDNN 7.6
Tensorflow 2.1.0

**Comment by ShrayBans** [comment]
Same issue here:
CUDA 10.1
CuDNN 7.6
Tensorflow 1.15

**Comment by Novarizark** [comment]
Same.
CUDA 10.1
CuDNN 7.6
Tensorflow 2.1.0

**Comment by chudyandrej** [comment]
Same.
CUDA 10.2
CuDNN 7.6.5.32
Tensorflow 2.1.0

**Comment by sommere** [comment]
Same:

* CUDA 11.0.1 RC
* cuDNN 8.0.2 RC2
* TensorFlow 2.2.0

**Comment by micbia** [comment]
Hi,

I had the same issue and I think I figured out a way around it. In my case, I am working on an HPC and I intalled keras on my /.local, whereas Tensorflow and CUDA are installed by the IT staff, anyway I encountered the same error above. I am using **Tensorflow==1.15.0** and **Keras==2.3.1**

I noticed that the message error:
- ValueError: To call `multi_gpu_model` with `gpus=2`, we expect the following devices to be available: ['/cpu:0', '/gpu:0', '/gpu:1'].
However this machine only has: ['/cpu:0', '/xla_cpu:0', '/xla_gpu:0', '/xla_gpu:1']. Try reducing `gpus`.

is located in the following file of keras, line 184:
- /home/.local/lib/python3.7/site-packages/keras/utils/multi_gpu_utils.py

I solved this, by replacing the **line 175** with the following:
- target_devices = ['/cpu:0'] + ['/gpu:%d' % i for i in target_gpu_ids] **(before)**
target_devices = ['/cpu:0'] + ['/xla_gpu:%d' % i for i in target_gpu_ids] **(after)**

Moreover, I modified the following keras file:
- /home/.local/lib/python3.7/site-packages/keras/backend/tensorflow_backend.py

so I replaced the line 510 with:
- return [x for x in _LOCAL_DEVICES if 'device:gpu' in x.lower()] **(before)**
return [x for x in _LOCAL_DEVICES if 'device:XLA_GPU' in x] **(after)**

Long story short only to say that apparently this is a bug of Keras and not of some environmental setup. 

After such modification my network was able to run with the xla_gpus, I hope this is somehow helpful.


Cheers,
Michele

**Comment by arb-git** [comment]
@micbia 
Tried your solution, but for me the files .../multi_gpu_utils.py and .../tensorflow_backend.py don't even exist.
I have tensorflow 2.3.1 though.
In general, is there any solution to tensorflow/keras being unable to use XLA-GPU that does not involve reinstalling CUDA+tf+cudnn?

**Comment by micbia** [comment]
@arb-git 
notice that path for the files I mentioned above are for **keras** module, If you are not using keras then I can suggest you to use the tf.keras. 

The location of the multi_gpu_utils.py module for TensorFlow2.3.0 (stable) is basically the same [tf.keras.utils.multi_gpu_model](https://www.tensorflow.org/api_docs/python/tf/keras/utils/multi_gpu_model), and the [line 172](https://github.com/tensorflow/tensorflow/blob/fcc4b966f1265f466e82617020af93670141b009/tensorflow/python/kerasurl/utils/multi_gpu_utils.py#L172) is to change as I suggested

In the case of tensorflow_backend.py it changes a little bit [tf.keras.backend.backend](https://www.tensorflow.org/api_docs/python/tf/keras/backend/backend) and the [line 783](https://github.com/tensorflow/tensorflow/blob/v2.3.1/tensorflow/python/keras/backend.py#L783) is to change, but by replacing:
-   return [x.name for x in _LOCAL_DEVICES if x.device_type == 'GPU'] (before)
-   return [x.name for x in _LOCAL_DEVICES if x.device_type == 'XLA_GPU'] (after)

I have not tried personally on TensorFlow2.3.0. It would be interesting if you could share at let us know if it works in your case with this changes.

Cheers,
Michele
